File: 21May/ass2.py
#!/usr/bin/env python3
import heapq, math
from itertools import count

class Node:
    def __init__(self, symbol=None, prob=0.0, left=None, right=None):
        self.symbol = symbol
        self.prob   = prob
        self.left   = left
        self.right  = right

def build_huffman_tree(symbols, probs):
    """Builds Huffman tree; returns its root."""
    heap = []
    counter = count()  # tie‑breaker
    for sym, p in zip(symbols, probs):
        heapq.heappush(heap, (p, next(counter), Node(symbol=sym, prob=p)))

    while len(heap) > 1:
        p1, _, n1 = heapq.heappop(heap)
        p2, _, n2 = heapq.heappop(heap)
        merged = Node(prob=p1 + p2, left=n1, right=n2)
        heapq.heappush(heap, (merged.prob, next(counter), merged))

    return heap[0][2]

def assign_codes(node, prefix="", codebook=None):
    """Recursively builds codebook {symbol: code}."""
    if codebook is None:
        codebook = {}
    # Leaf?
    if node.left is None and node.right is None:
        # If only one symbol, give it code "0"
        codebook[node.symbol] = prefix or "0"
    else:
        assign_codes(node.left,  prefix + "0", codebook)
        assign_codes(node.right, prefix + "1", codebook)
    return codebook

def main():
    # 1) Read inputs
    M = int(input("Enter number of symbols M: "))
    symbols, probs = [], []
    print("Enter each symbol and its probability (e.g. A 0.25):")
    for _ in range(M):
        s, p = input().split()
        symbols.append(s)
        probs.append(float(p))

    # 2) Build tree and codebook
    root  = build_huffman_tree(symbols, probs)
    codes = assign_codes(root)

    # 3) Print symbol, prob, code
    print("\nSymbol\tProb\tCode")
    for s, p in zip(symbols, probs):
        print(f"{s}\t{p:.4f}\t{codes[s]}")

    # 4) Compute average length and entropy
    L = sum(p * len(codes[s]) for s, p in zip(symbols, probs))
    H = -sum(p * math.log2(p) for p in probs if p > 0)

    print(f"\nAverage code length L = {L:.4f}")
    print(f"Entropy H            = {H:.5f} bits/symbol")

if __name__ == "__main__":
    main()

File: 21May/ass2.cpp
// ass2.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <map>
#include <string>
#include <cmath>
#include <algorithm>   // for completeness, though not strictly needed below

using namespace std;

// 1) Huffman tree node
struct Node {
    string symbol;
    double prob;
    Node *left, *right;
    Node(const string& s, double p)
      : symbol(s), prob(p), left(nullptr), right(nullptr) {}
};

// 2) Comparator for min‑heap (lowest prob at top)
struct Compare {
    bool operator()(const Node* a, const Node* b) const {
        return a->prob > b->prob;
    }
};

// 3) Recursively walk the Huffman tree and build codewords
void assignCodes(Node* root, const string& prefix, map<string,string>& codes) {
    if (!root) return;
    // Leaf node
    if (!root->left && !root->right) {
        codes[root->symbol] = prefix;
    } else {
        assignCodes(root->left,  prefix + "0", codes);
        assignCodes(root->right, prefix + "1", codes);
    }
}

int main() {
    int M;
    cout << "Enter number of symbols M: ";
    if (!(cin >> M) || M <= 0) {
        cerr << "Invalid M\n"; 
        return 1;
    }

    vector<string> symbols(M);
    vector<double> probs(M);
    cout << "Enter each symbol and its probability (e.g. A 0.25):\n";
    for (int i = 0; i < M; i++) {
        cin >> symbols[i] >> probs[i];
        if (!cin || probs[i] < 0) {
            cerr << "Bad input\n";
            return 1;
        }
    }

    // Build priority queue of leaf nodes
    priority_queue<Node*, vector<Node*>, Compare> pq;
    for (int i = 0; i < M; i++) {
        pq.push(new Node(symbols[i], probs[i]));
    }

    // Merge until one tree remains
    while (pq.size() > 1) {
        Node* x = pq.top(); pq.pop();
        Node* y = pq.top(); pq.pop();
        Node* z = new Node("", x->prob + y->prob);
        z->left  = x;
        z->right = y;
        pq.push(z);
    }
    Node* root = pq.top();

    // Generate codes
    map<string,string> codes;
    assignCodes(root, "", codes);

    // Print symbol, probability, code
    cout << "\nSymbol\tProb\tCode\n";
    for (int i = 0; i < M; i++) {
        cout 
          << symbols[i] << '\t'
          << probs[i]   << "\t"
          << codes[symbols[i]] 
          << "\n";
    }

    // Compute average length L and entropy H
    double L = 0.0, H = 0.0;
    for (int i = 0; i < M; i++) {
        double p = probs[i];
        int    ℓ = codes[symbols[i]].length();
        L += p * ℓ;
        if (p > 0) H -= p * log2(p);
    }

    cout << "\nAverage code length L = " << L << "\n";
    cout << "Entropy H            = " << H << " bits/symbol\n";

    // (Optional) clean up allocated nodes omitted for brevity

    return 0;
}
